apm-agent-plugins/apm-micrometer-plugin/src/main/java/co/elastic/apm/agent/micrometer/MicrometerMetricsReporter.java

```diff
@@ -0,0 +1,112 @@
+import co.elastic.apm.agent.impl.Tracer;
+import co.elastic.apm.agent.report.Reporter;
+import co.elastic.apm.agent.report.ReporterConfiguration;
+import co.elastic.apm.agent.sdk.weakmap.WeakMapSupplier;
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentSet;
+import com.dslplatform.json.DslJson;
+import com.dslplatform.json.JsonWriter;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.composite.CompositeMeterRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+public class MicrometerMetricsReporter implements Runnable {
+
+    private static final Logger logger = LoggerFactory.getLogger(MicrometerMetricsReporter.class);
+
+    private final WeakConcurrentSet<MeterRegistry> meterRegistries = WeakMapSupplier.createSet();
+    private final StringBuilder replaceBuilder = new StringBuilder();
+    private final JsonWriter jsonWriter = new DslJson<>(new DslJson.Settings<>()).newWriter();
+    private final Reporter reporter;
+    private final ElasticApmTracer tracer;
+    private boolean scheduledReporting = false;
+
+    public MicrometerMetricsReporter(ElasticApmTracer tracer) {
+        this.tracer = tracer;
+        this.reporter = tracer.getReporter();
+    }
+
+    public void registerMeterRegistry(MeterRegistry meterRegistry) {
+        if (meterRegistry instanceof CompositeMeterRegistry) {
+            return;
+        }
+        boolean added = meterRegistries.add(meterRegistry);
+        if (added) {
+            logger.info("Registering Micrometer MeterRegistry: {}", meterRegistry);
+            scheduleReporting();
+        }
+    }
+
+    private synchronized void scheduleReporting() {
+        if (scheduledReporting) {
+            return;
+        }
+        scheduledReporting = true;
+        long metricsIntervalMs = tracer.getConfig(ReporterConfiguration.class).getMetricsIntervalMs();
+        if (metricsIntervalMs > 0) {
+            // called for every class loader that loaded micrometer
+            tracer.getSharedSingleThreadedPool().scheduleAtFixedRate(this, metricsIntervalMs, metricsIntervalMs, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    // guaranteed to be invoked by a single thread
+    @Override
+    public void run() {
+        if (tracer.getState() != Tracer.TracerState.RUNNING) {
+            return;
+        }
+        final long timestamp = System.currentTimeMillis() * 1000;
+        MeterMapConsumer meterConsumer = new MeterMapConsumer();
+        for (MeterRegistry registry : meterRegistries) {
+            registry.forEachMeter(meterConsumer);
+        }
+        logger.debug("Reporting {} meters", meterConsumer.meters.size());
+        MicrometerMeterRegistrySerializer.serialize(meterConsumer.meters, timestamp, replaceBuilder, jsonWriter);
```

Comment:
One reason is to make sure that metrics are serialized at a more consistent rate. Currently, the serialization is delayed until the event is consumed from the disruptor. That's something we have thought about doing for the our internal metrics registry as well.
Another reason is that I didn't want to add a micrometer specific event type for the disruptor. But we could introduce an interface that allows writing to the `OutputStream`.

Reply:
Yes, we discussed that and I actually think this approach is superior, it's just the inconsistency between the two registries. I would change the internal metrics to be done in this way as well (it even makes more sense there where we flip phase on the disruptor thread). With this new `byte[]` reporting event and API, it should be an easy change. Doesn't have to be in this PR though.

Reply:
do you think we should have a way to recycle these byte arrays or, as reporting happens at a fixed rate that's not dependant on the throughput of the app, it's fine to just allocate?

Reply:
As long as it is used for metrics, meaning allocated once per second or more, garbage collection will not be an issue. Keeping such an array we always hold the biggest array that was used so far and cannot be reclaimed (and now when we use if for Micrometer I guess it may have spikes). 
Having said that, this is what the `JsonWriter` is doing already, so it does seem wasteful to duplicate. So it seems whatever we decide (allocate or recycle) should be done should be done through `JsonWriter` objects, eventually relying on the `com.dslplatform.json.JsonWriter#toStream` method to avoid additional allocation. WDYT?

Reply:
I'd expect the total size of the metrics to be relatively stable. `JsonWriter#toStream` sounds tempting but it would imply that you can't do another report before the bytes of the previous report have been written out. This can be problematic if APM Server is down for a couple of minutes. By copying the bytes via `JsonWriter#toByteArray`, all the data from the period where APM Server was down will be sent when it's back up again.

Reply:
We can use that with your suggestion for an interface that writes to the `OutputStream` (wrapping the `JsonWriter`), which will be added to the `ReportingEvent` instead of `byte[]`

Reply:
> I'd expect the total size of the metrics to be relatively stable

Based on your docs, people may use it so report some high-cardinality metric keys

Reply:
Yup, high cardinality keys can be problematic but still, I think that the size of one report vs another one would be relatively stable.

> What I was suggesting is to either create a new one every time (instead of byte[]) or recycle (instead of byte[])

I see what you're saying. One issue I see with the approach of using a new byte[] every time is that it'll have to discard lots of arrays when it resizes from the initial buffer size.

Reply:
> One issue I see with the approach of using a new byte[] every time is that it'll have to discard lots of arrays when it resizes from the initial buffer size.

I think the eventual allocation and copy of the entire byte array is not necessary in any case, so I suggest one of two options:
1. Allocate a `JsonWriter` every time, using the size from last time as the initial capacity
2. Recycle `JsonWriter`s
Would that work?

Reply:
Thanks for switching the internal metrics to be the same! ❤️

Reply:
I went with option 1. I'm using the size of the previous serialization plus a buffer of 25%. This is to avoid situations where we have to resize just because the actual size requirement is slightly higher which may be just due to the fact that on one report the gauges have more floating points (like 1.0 vs 1.00000001).

Reply:
Option 2 can be implemented later on top of option 1 if needed.